django/db/models/sql/compiler.py

```
@@ -1846,15 +1846,31 @@ def pre_sql_setup(self):
         # Now we adjust the current query: reset the where clause and get rid
         # of all the tables we don't need (since they're in the sub-select).
         self.query.clear_where()
+        self.query.related_ids = {}
         if self.query.related_updates or must_pre_select:
             # Either we're using the idents in multiple update queries (so
             # don't want them to change), or the db backend doesn't support
             # selecting from the updating table (e.g. MySQL).
             idents = []
-            for rows in query.get_compiler(self.using).execute_sql(MULTI):
+            compiler = query.get_compiler(self.using)
+            for rows in compiler.execute_sql(MULTI):
                 idents.extend(r[0] for r in rows)
             self.query.add_filter("pk__in", idents)
-            self.query.related_ids = idents
+            concrete_model = self.query.get_meta().concrete_model
+            self.query.related_ids[concrete_model] = idents
+            if self.query.related_updates:
+                for model in self.query.related_updates:
+                    parent_link = self.query.get_meta().get_ancestor_link(model)
+                    if parent_link is None:
+                        continue
```

[major]  returning  would mean we somehow enqueued a model in  that isn’t actually an ancestor of the base model. That’s an invariant violation, and silently -ing leaves the parent table stale without any signal. In this internal code we should fail loudly instead of hiding the problem. Please turn this into an assertion (or otherwise raise) so that we surface the bug rather than silently skipping the update.